clang/lib/Sema/SemaChecking.cpp

```diff
@@ -19812,11 +19821,13 @@ bool Sema::SemaBuiltinElementwiseTernaryMath(CallExpr *TheCall) {
     Args[I] = Converted.get();
   }
 
-  int ArgOrdinal = 1;
-  for (Expr *Arg : Args) {
-    if (checkFPMathBuiltinElementType(*this, Arg->getBeginLoc(), Arg->getType(),
-                                      ArgOrdinal++))
-      return true;
+  if (enforceFloatingPointCheck) {
+    int ArgOrdinal = 1;
+    for (Expr *Arg : Args) {
+      if (checkFPMathBuiltinElementType(*this, Arg->getBeginLoc(),
+                                        Arg->getType(), ArgOrdinal++))
```

Comment:
Should this call some other element type check instead in this case? I initially thought `checkMathBuiltinElementType` would make sense, but from a quick glance that seems to be tied to matrix types for some reason.

Reply:
mad is suppose to operate on matrix types, we just don't have them yet.

Reply:
I was mostly commenting on the jump from "math builtin" to "matrix type" as far as naming goes, but I don't spend a lot of time in Sema. My point is that I think we may still want some check here, I'm just not sure what without looking at it in more detail.

Reply:
So looking at `checkMathBuiltinElementType` it use of `isValidElementType` works for more than just matrix types  so  i think it is fine to use.  SO to be clear below is what you would like to see?

```cpp
if (CheckForFloatArgs) {
    int ArgOrdinal = 1;
    for (Expr *Arg : Args) {
      if (checkFPMathBuiltinElementType(*this, Arg->getBeginLoc(),
                                        Arg->getType(), ArgOrdinal++))
        return true;
    }
  } else  {
    int ArgOrdinal = 1;
    for (Expr *Arg : Args) {
      if (checkMathBuiltinElementType(*this, Arg->getBeginLoc(),
                                        Arg->getType(), ArgOrdinal++))
        return true;
}
```

Reply:
The impact is it just changed the error when we use bool

Before change:
`// expected-error@-1 {{arguments are of different types ('double' vs 'bool')}}`
After change:
`// expected-error@-1 {{3rd argument must be a vector, integer or floating point type (was 'bool')}}`